common/src/chainstate/utxo.rs

```diff
@@ -0,0 +1,462 @@
+    pub fn new(output: TxOutput, is_block_reward: bool, height: BlockHeight) -> Self {
+        Self {
+            output,
+            is_block_reward,
+            height,
+        }
+    }
+
+    pub fn is_block_reward(&self) -> bool {
+        self.is_block_reward
+    }
+
+    pub fn height(&self) -> BlockHeight {
+        self.height
+    }
+
+    pub fn output(&self) -> &TxOutput {
+        &self.output
+    }
+
+    pub fn block_reward(&mut self, value: bool) {
+        self.is_block_reward = value;
+    }
+
+    pub fn set_height(&mut self, value: BlockHeight) {
+        self.height = value;
+    }
+}
+
+#[derive(Debug, Clone, Eq, PartialEq, Encode, Decode)]
+pub struct UtxoEntry {
+    /// None when utxo has been spent.
+    utxo: Option<Utxo>,
+    /// The utxo entry is dirty when this version is different from the parent.
+    is_dirty: bool,
+    /// The utxo entry is fresh when the parent does not have this utxo
+    is_fresh: bool,
+}
+
+impl UtxoEntry {
+    pub fn new(utxo: Utxo, is_fresh: bool, is_dirty: bool) -> Self {
+        Self {
+            utxo: Some(utxo),
+            is_dirty,
+            is_fresh,
+        }
+    }
+
+    pub(in crate::chainstate) fn new_spent(is_fresh: bool, is_dirty: bool) -> Self {
+        Self {
+            utxo: None,
+            is_dirty,
+            is_fresh,
+        }
+    }
+
+    fn utxo_mut(&mut self) -> Option<&mut Utxo> {
+        self.utxo.as_mut()
+    }
+
+    pub fn is_dirty(&self) -> bool {
+        self.is_dirty
+    }
+
+    pub fn is_fresh(&self) -> bool {
+        self.is_fresh
+    }
+
+    pub fn is_spent(&self) -> bool {
+        self.utxo.is_none()
+    }
+
+    pub fn utxo(&self) -> Option<Utxo> {
+        self.utxo.clone()
+    }
+}
+
+pub trait UtxosView {
+    /// Retrieves utxo.
+    fn get_utxo(&self, outpoint: &OutPoint) -> Option<Utxo>;
+
+    /// Checks whether outpoint is unspent.
+    fn have_utxo(&self, outpoint: &OutPoint) -> bool;
+
+    /// Retrieves the block hash of the best block in this view
+    fn get_best_block_hash(&self) -> Option<H256>;
+
+    /// Estimated size of the whole view (0 if not implemented)
+    fn estimated_size(&self) -> usize;
+
+    /// Performs bulk modification
+    fn batch_write(
+        &mut self,
+        utxos: HashMap<OutPointKey, UtxoEntry>,
+        block_hash: H256,
+    ) -> Result<(), Error>;
+
+    fn derive_cache(&self) -> UtxosCache;
+}
+
+pub fn flush_to_base<T: UtxosView>(
+    cache: UtxosCache,
+    block_hash: H256,
+    base: &mut T,
+) -> Result<(), Error> {
+    base.batch_write(cache.utxos, block_hash)
+}
+
+#[derive(Clone, Default)]
+pub struct UtxosCache<'a> {
+    parent: Option<&'a dyn UtxosView>,
+    current_block_hash: Option<H256>,
+    pub(in crate::chainstate) utxos: HashMap<OutPointKey, UtxoEntry>,
+    memory_usage: usize,
+}
+
+impl<'a> UtxosCache<'a> {
+    /// returns a copy of the UtxoEntry, given the outpoint.
+    fn get_utxo_entry(&self, outpoint: &OutPoint) -> Option<UtxoEntry> {
```

Comment:
I'm really wondering whether there's a difference between this and `UtxosView::get_utxo()`.

Reply:
the `get_utxo_entry` is needed to identify whether the utxo is: 
dirty -> if the utxo version is different from the parent.
fresh -> when the utxo only exists in that view.

the `get_utxo_entry` is needed for `spend_utxo`.

Reply:
This method is private and isn't for the user.